Share a single etcd3 client logger across all clients

negz opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?

<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

/kind bug

#### What this PR does / why we need it:

Currently the API server creates one etcd client per CRD. If clients aren't provided a logger they'll each create their own. These loggers can account for ~20% of API server memory consumption on a cluster with hundreds of CRDs.

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes https://github.com/kubernetes/kubernetes/issues/111476

#### Special notes for your reviewer:

With this single logger and ~1,900 CRDs loaded I'm seeing the API server using from 5.5 to 6.5 GiB RSS as garbage collection sawtooths. That's a saving of at least 1GiB.

Profile no longer shows the logger:

![profile-with-one-logger](https://user-images.githubusercontent.com/1049349/181673236-f9247ada-e0d7-4d53-9f40-a1770142a41b.png)

Note that I don't really know how to make this logger actually log anything so it's hard to confirm whether this change has any performance or observability impact.

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

negz committed:
Disable the etcd3 client logger

This logger is responsible for 20% of the API server's memory usage when
many CRDs are installed. See the below issue for more context.

https://github.com/kubernetes/kubernetes/issues/111476

Signed-off-by: Nic Cope <nicc@rk0n.org>

leilajal commented:
/cc @apelisse @jpbetz 
/triage accepted

jpbetz commented:
> sure how viable completely disabling it is; this PR is intended more to start a discussion and to demonstrate what I did in https://github.com/kubernetes/kubernetes/issues/111476 to disable the logger

re: viability: Sounds like you're already thinking in this direction-- I'm relatively certain there is information being logged that would be missed by cluster administrators when diagnosing etcd issue.

Any idea what the current log level is and what volume of data was being logged in the 1k+ CRD case?

negz commented:
> Any idea what the current log level is and what volume of data was being logged in the 1k+ CRD case?

I don't see anything being logged, but I just have a toy kind cluster with the CRDs loaded and no controllers or real clients. I imagine there might be interesting logs on a real cluster.

Per https://github.com/kubernetes/kubernetes/issues/111476#issuecomment-1198803133 this is almost certainly happening because we have one etcd client (and thus one logger) per CRD installed. I imagine we should probably share a client or pool of clients to etcd, but I'd love to get a fix for this memory usage into 1.25 and I believe the code freeze is days away.

Perhaps a reasonable quick fix for this logger issue would be to create one logger and pass it into all newly created etcd clients?

negz committed:
Share a single etcd3 client logger across all clients

Currently the API server creates one etcd client per CRD. If clients
aren't provided a logger they'll each create their own. These loggers
can account for ~20% of API server memory consumption on a cluster with
hundreds of CRDs.

Signed-off-by: Nic Cope <nicc@rk0n.org>

negz commented:
Just pushed a commit and edited the PR description to have clients share one logger rather than each creating their own.

sttts commented on staging/src/k8s.io/apiserver/pkg/storage/storagebackend/factory/etcd3.go:
```diff
@@ -63,13 +66,27 @@ const (
+	if err != nil {
+		l = zap.NewNop()
+	}
+	etcd3ClientLogger = l
```
generally this looks fine as a stop-the-bleeding solution. Only open question as discussed on Slack: somewhere is an integration of this logger into klog if I remember right. Would double check that this change does not influence that. Probably it won't.

sttts replied:
cc @wojtek-t

negz replied:
FWIW this is where the logger was previously being setup - https://github.com/etcd-io/etcd/blob/v3.5.4/client/v3/client.go#L374

negz replied:
...which makes me think I should probably give the logger a name as they do in the original code and potentially honor the debug level env var too.

negz replied:
Done. I did push the `ETCD_CLIENT_DEBUG` logic in a separate commit just because it looks strange to me - unless I'm missing something it returns the info (default) debugging level no matter what the env var is set to.

sttts commented:
/ok-to-test

aojea commented on staging/src/k8s.io/apiserver/pkg/storage/storagebackend/factory/etcd3.go:
```diff
@@ -63,13 +66,27 @@ const (
+// creating an etcd client for each CRD. We need to pass each etcd client a
+// logger or each client will create its own, which comes with a significant
+// memory cost (around 20% of the API server's memory when hundreds of CRDs are
+// present). The correct fix here is to not create a client per CRD. See
```
we create a client per CRD and API type, but sharing the client doesn't work https://github.com/kubernetes/kubernetes/pull/105826

I couldn't investigate in detail why this happens

jpbetz commented:
> Perhaps a reasonable quick fix for this logger issue would be to create one logger and pass it into all newly created etcd clients?

I like this idea. If we can pool etcd clients event better, but if we are unable to do that, or if we are able but it is going to take a long time, this might be worth doing.

negz committed:
Give etcd client logger a name

Logic copied from https://github.com/etcd-io/etcd/blob/v3.5.4/client/v3/client.go#L374

Signed-off-by: Nic Cope <nicc@rk0n.org>

negz committed:
Copy etcd client debug level logic from upstream

Replicated from https://github.com/etcd-io/etcd/blob/v3.5.4/client/v3/logger.go#L47

The logic of this function doesn't make a lot of sense to me, but
copying it will avoid any behaviour change.

Signed-off-by: Nic Cope <nicc@rk0n.org>

negz commented:
> If we can pool etcd clients event better, but if we are unable to do that, or if we are able but it is going to take a long time, this might be worth doing.

I'm happy to take a shot at that in a separate PR, but I'd prefer to get this merged first given how close we are to 1.25 freeze, just to make sure we get the memory reduction with 1.25 either way.

jpbetz commented:
The way the initialization of the default etcd logger is recreated here to have a shared logger looks right to me.
/lgtm

negz commented:
/test pull-kubernetes-e2e-kind-ipv6

negz commented:
> I'm happy to take a shot at that in a separate PR

Draft started in https://github.com/kubernetes/kubernetes/pull/111559

negz commented:
I don't think I'll personally be confident enough in https://github.com/kubernetes/kubernetes/pull/111559 to get it done by the freeze tomorrow. Do folks feel like this one is ready to merge as a stop-gap?

jqmichael commented:
Related discussion about one etcd client per API Kind. 

https://kubernetes.slack.com/archives/C0EG7JC6T/p1618439560231400

> [lavalamp](https://app.slack.com/team/U09R0TW1F)
  [1 year ago](https://kubernetes.slack.com/archives/C0EG7JC6T/p1618441955232000?thread_ts=1618439560.231400&cid=C0EG7JC6T)
Yeah that sounds like what it currently does. What it should do is make one client per database (recall that you can put different collections in different etcd databases). 
[@jingyih](https://kubernetes.slack.com/team/UBALVKL9K)
 might have a patch laying around that fixes this.

negz commented:
> What it should do is make one client per database (recall that you can put different collections in different etcd databases).

@jqmichael that sounds similar to what I've done in https://github.com/kubernetes/kubernetes/pull/111559 - I believe in this case one transport is roughly equivalent to one etcd database?

lavalamp commented:
/lgtm
/approve
